### PR TITLE
Regions Plugins - prevent Click when creating / updating region

### PR DESCRIPTION
**Pull request by ArTiSTiX**
When creating / updating regions, cursor seeks to mouseup position.
Such behaviour should be prevented, and must be part of application logic (e.g. by listening to `region-updated` or `region-created`, seekTo start of region might be useful).
Thus, interactions with regions plugin must be ignored by drawer.

# Fixes
* Prevent once following click event after updating or creating a region with a preventClick util function
* Added mouse up listener on body when creating region to stop resize if cursor leaves waveform
* Do not fire region update-end event if nothing changed

# Notes:
The only way to prevent click events to "bubble" after a `mouseup` is to capture following click event from body (or a parent to the listener), and stop its propagation.

So i added a preventClick util to expose it to future plugins. It should work when called from any mouseDown/mouseMove/mouseUp events

Reference/Idea: https://stackoverflow.com/questions/33364540/prevent-click-event-from-mousedown

# Related Issues:
https://github.com/katspaugh/wavesurfer.js/issues/1064 - Bad region dragging behaviour with cursor
https://github.com/katspaugh/wavesurfer.js/issues/572 - seekTo and seek event fires when creating a region

# Testing
Only tested on Chrome 63

### CODE REVIEW

**Comment by thijstriemstra** [comment]
@roboswank can you take a look at this?

**Comment by thijstriemstra** [comment]
@ArTiSTiX this PR has some conflicts now, can you take a look at that?

**Comment by ArTiSTiX** [comment]
@thijstriemstra I will have a look tomorrow, and test it with all new changes from 2.0.1 (the version i use).
I think the conflict can be resolved by accepting both additions (in any order, it's just a variable declaration).

Thanks for considering this merge.

**Comment by ArTiSTiX** [comment]
I just noticed mouse events now ends as soon as the cursor leaves the container.
It might be a recent planned behaviour, but I think it's more user-friendly to let him drag off the wave-form.

Didn't had time to inspect concerned changes/issues today.

**Comment by thijstriemstra** [comment]
> I just noticed mouse events now ends as soon as the cursor leaves the container.
It might be a recent planned behaviour, but I think it's more user-friendly to let him drag off the wave-form.

@roboswank can you give some feedback on that?

**Comment by ArTiSTiX** [comment]
@thijstriemstra @roboswank Sorry, i remembered wrongly. 
✅ Leaving container doesn't end region dragging, but outer moves doesn't update (i thought it used to receive mousemoves but it never worked like that, however it could be an improvement).
✅ Re-entering still drags
✅ mouseup/touchend ends dragging even if out of container (i fixed it in this PR, wasn't working when creating regions)
❓ Moving mouse could drag region out of container by creating the move listener on document instead of the wrapper

**Comment by thijstriemstra** [comment]
@ArTiSTiX I do not use the regions plugin, nor do I pretend to understand it's code, so if you can make the necessary adjustments in this pull request, we can take a look.

**Comment by ArTiSTiX** [comment]
I prefer keeping this PR light, only fixing related issues and not implementing new behaviour.
I will file an dedicated issue (open to discussion) about the possibility of dragging out of container to improve user experience.
I plan to work on other improvements requested on regions plugins (e.g. non-overlapping regions).

File: src/plugin/regions.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ class Region {
         this.wavesurfer = ws;
         this.wrapper = ws.drawer.wrapper;
         this.style = ws.util.style;
+        this.util = ws.util;
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: can you swap line 13 and 14 and use this.style = this.util.style?

File: src/util/prevent-click.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+function preventClickHandler(e) {
```
Comment (thijstriemstra) [inline src/util/prevent-click.js]: Can you add a docstring to this module? e.g.

```
/**
 * Description here
 *
 * @return {function}
 */
```

**Comment by johndebord** [comment]
On mobile, if a region is moved or adjusted, the user has to double-tap to get anything to work.
What's the recommended work-around for this?

**Pull request merged**